File: Java/src/main/java/com/leanrose/ElectricRose.java
package com.leanrose;

class ElectricRose {
    Battery[] batteries;

    public ElectricRose(Battery[] batteries) {
        this.batteries = batteries;
    }

    public void updateCharge() {
        for (int i = 0; i < batteries.length; i++) {
            if (batteries[i].charge >= 0.1) {
                batteries[i].charge = Math.max(0, batteries[i].charge - 0.1);
            }

            if (batteries[i].dry) {
                if (batteries[i].charge >= 0.1) {
                    batteries[i].charge = Math.max(0, batteries[i].charge - 0.1);
                }
            }
        }
    }
}

File: CSharp/ElectricRoseBatteryTracker/ElectricRose.cs
using ElectrictRoseBatteryTracker.Models;

namespace ElectrictRoseBatteryTracker;

public class ElectricRose
{
    private Battery[] batteries;

    public ElectricRose(Battery[] batteries)
    {
        this.batteries = batteries;
    }

    public void UpdateCharge()
    {
        foreach (var battery in batteries)
        {
            if (battery.Charge >= 0.1)
            {
                battery.Charge = Math.Max(0, battery.Charge - 0.01);
            }

            if (battery.Dry)
            {
                if (battery.Charge >= 0.1)
                {
                    battery.Charge = Math.Max(0, battery.Charge - 0.01);
                }
            }
        }
    }
}